### PR TITLE
perf!: FindLatestAuthorizationModelID changes to FindLatestAuthorizationModel

### PR DESCRIPTION
**Pull request by adriantam**
## Description
Change storage interface FindLatestAuthorizationModelID to FindLatestAuthorizationModel, this allows optimization for type system resolver if model ID is not specified by requiring single DB lookup instead of two.

Note that the interface for storage AuthorizationModelReadBackend is modified.

## References

Close https://github.com/openfga/openfga/issues/1386

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/storage/sqlcommon/sqlcommon.go [hunk 1]
```diff
@@ -549,6 +533,47 @@ func ReadAuthorizationModel(
 	}, nil
 }
 
+// ReadLatestAuthorizationModel reads the latest authorization model corresponding to the store
+func ReadLatestAuthorizationModel(
```
Comment (jon-whit) [inline pkg/storage/sqlcommon/sqlcommon.go:536]: Can we please verify this is getting served from an index and avoiding a larger scan?

If we create several tens of thousands of models and write them to our store does the OrderBy followed by Limit degrade with more models. I would suspect it does. Let's confirm/deny this.

File: pkg/storage/mysql/mysql.go [hunk 2]
```diff
@@ -434,25 +434,12 @@ func (m *MySQL) ReadAuthorizationModels(
...
-		OrderBy("authorization_model_id desc").
-		Limit(1).
-		QueryRowContext(ctx).
-		Scan(&modelID)
-	if err != nil {
-		return "", sqlcommon.HandleSQLError(err)
-	}
-
-	return modelID, nil
+	return sqlcommon.ReadLatestAuthorizationModel(ctx, sqlcommon.NewDBInfo(m.db, m.stbl, "NOW()"), store)
```
Comment (miparnisari) [inline pkg/storage/mysql/mysql.go:442]:
```suggestion
	return sqlcommon.FindLatestAuthorizationModel(ctx, sqlcommon.NewDBInfo(m.db, m.stbl, "NOW()"), store)
```

File: pkg/storage/storagewrappers/caching.go [hunk 3]
```diff
@@ -52,14 +52,14 @@ func (c *cachedOpenFGADatastore) ReadAuthorizationModel(ctx context.Context, sto
 }
 
 // FindLatestAuthorizationModelID returns the last model `id` written for a store.
-func (c *cachedOpenFGADatastore) FindLatestAuthorizationModelID(ctx context.Context, storeID string) (string, error) {
-	v, err, _ := c.lookupGroup.Do(fmt.Sprintf("FindLatestAuthorizationModelID:%s", storeID), func() (interface{}, error) {
-		return c.OpenFGADatastore.FindLatestAuthorizationModelID(ctx, storeID)
+func (c *cachedOpenFGADatastore) FindLatestAuthorizationModelID(ctx context.Context, storeID string) (*openfgav1.AuthorizationModel, error) {
```
Comment (miparnisari) [inline pkg/storage/storagewrappers/caching.go:55]: Ohh this is such a sneaky bug..

your change:
image

but we should update the name of the method too:

image

File: pkg/storage/storagewrappers/caching_test.go [hunk 4]
```diff
@@ -54,37 +50,9 @@ func TestReadAuthorizationModel(t *testing.T) {
...
-				return err
-			}
-			if id != expectedModelID {
-				return fmt.Errorf("expected model ID %s, actual %s", expectedModelID, id)
-			}
-			return nil
-		})
-	}
-	err := wg.Wait()
+	// ensure find latest authorization model will get hte latest model
```
Comment (miparnisari) [inline pkg/storage/storagewrappers/caching_test.go:54]: This assertion isn't enough. we should change

```go
memoryBackend := memory.New()
cachingBackend := NewCachedOpenFGADatastore(memoryBackend, 5)
```

to 

```go
mockController := gomock.NewController(t)
defer mockController.Finish()
mockDatastore := mocks.NewMockOpenFGADatastore(mockController)
mockDatastore.EXPECT().FindLatestAuthorizationModel(gomock.Any(), gomock.Any()).Times(1)
cachingBackend := NewCachedOpenFGADatastore(mockDatastore, 5)
```

note the `Times(1)` :)

**Review by miparnisari — CHANGES_REQUESTED** [review]

File: pkg/storage/mysql/mysql.go (hunk 2 above)
Comment (jon-whit) [inline pkg/storage/mysql/mysql.go:442]: nit (non-blocking) - why do we create a new `sqlcommon.NewDBInfo` in various places. This is a structure that only needs to be created once upon construction of the MySQL adapter. This should belong in the constructor IMO.

File: pkg/storage/postgres/postgres.go [hunk 5]
```diff
@@ -431,25 +431,12 @@ func (p *Postgres) ReadAuthorizationModels(ctx context.Context, store string, op
 	return models, token, nil
 }
 
-// FindLatestAuthorizationModelID see [storage.AuthorizationModelReadBackend].FindLatestAuthorizationModelID.
-func (p *Postgres) FindLatestAuthorizationModelID(ctx context.Context, store string) (string, error) {
-	ctx, span := tracer.Start(ctx, "postgres.FindLatestAuthorizationModelID")
+// FindLatestAuthorizationModel see [storage.AuthorizationModelReadBackend].FindLatestAuthorizationModelID.
```
Comment (jon-whit) [inline pkg/storage/postgres/postgres.go:434]: This comment and the others in this PR need to update the `FindLatestAuthorizationModelID` referenced in this sentence.

Please do a project-wide find/replace on `FindLatestAuthorizationModelID`.

File: pkg/storage/postgres/postgres.go [hunk 6]
```diff
@@ -431,25 +431,12 @@ func (p *Postgres) ReadAuthorizationModels(ctx context.Context, store string, op
...
+func (p *Postgres) FindLatestAuthorizationModel(ctx context.Context, store string) (*openfgav1.AuthorizationModel, error) {
+	ctx, span := tracer.Start(ctx, "mysql.FindLatestAuthorizationModel")
```
Comment (jon-whit) [inline pkg/storage/postgres/postgres.go:436]:
```suggestion
	ctx, span := tracer.Start(ctx, "postgres.FindLatestAuthorizationModel")
```

File: pkg/storage/storagewrappers/caching_test.go (hunk 4 above)
Comment (adriantam) [reply pkg/storage/storagewrappers/caching_test.go:54]: FindLatestAuthorizationModel is not serving from cache though

File: pkg/storage/sqlcommon/sqlcommon.go (hunk 1 above)
Comment (adriantam) [reply pkg/storage/sqlcommon/sqlcommon.go:536]: In a system with many stores, we see the following in Postgres

```
postgres=# explain select authorization_model_id, schema_version, type, type_definition, serialized_protobuf from authorization_model where store = '01HQBMGR6AV7H6MT4ZVK3CARR4' order by authorization_model_id desc limit 1;
                                                      QUERY PLAN
----------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.15..6.17 rows=1 width=160)
   ->  Index Scan Backward using authorization_model_pkey on authorization_model  (cost=0.15..12.18 rows=2 width=160)
         Index Cond: (store = '01HQBMGR6AV7H6MT4ZVK3CARR4'::text)
(3 rows)

```

File: pkg/storage/storagewrappers/caching.go (hunk 3 above)
Comment (adriantam) [reply pkg/storage/storagewrappers/caching.go]: fixed

File: pkg/storage/postgres/postgres.go (hunk 6 above)
Comment (adriantam) [reply pkg/storage/postgres/postgres.go]: fixed

File: pkg/storage/mysql/mysql.go (hunk 2 above)
Comment (adriantam) [reply pkg/storage/mysql/mysql.go]: updated

File: pkg/storage/storagewrappers/caching_test.go (hunk 4 above)
Comment (miparnisari) [reply pkg/storage/storagewrappers/caching_test.go:54]: But this is a unit test for the cache decoration

File: pkg/storage/storagewrappers/caching_test.go [hunk 7]
```diff
@@ -54,37 +58,9 @@ func TestReadAuthorizationModel(t *testing.T) {
 	gotModel, err = cachingBackend.ReadAuthorizationModel(ctx, storeID, model.Id)
 	require.NoError(t, err)
 	require.Equal(t, model, gotModel)
-}
-
-func TestSingleFlightFindLatestAuthorizationModelID(t *testing.T) {
```
Comment (miparnisari) [inline pkg/storage/storagewrappers/caching_test.go:59]: why remove this test? we are still using singleflight 🤔

File: pkg/storage/storagewrappers/caching_test.go [hunk 8]
```diff
@@ -3,26 +3,25 @@ package storagewrappers
...
 func TestReadAuthorizationModel(t *testing.T) {
 	ctx := context.Background()
-	memoryBackend := memory.New()
-	cachingBackend := NewCachedOpenFGADatastore(memoryBackend, 5)
-	defer cachingBackend.Close()
+	mockController := gomock.NewController(t)
+	mockController.Finish()
 
+	mockDatastore := mocks.NewMockOpenFGADatastore(mockController)
+	cachingBackend := NewCachedOpenFGADatastore(mockDatastore, 5)
```
Comment (miparnisari) [inline pkg/storage/storagewrappers/caching_test.go:24]:
```suggestion
	t.Cleanup(func() {
		goleak.VerifyNone(t)
	})
	ctx := context.Background()
	mockController := gomock.NewController(t)
	mockController.Finish()

	mockDatastore := mocks.NewMockOpenFGADatastore(mockController)
	cachingBackend := NewCachedOpenFGADatastore(mockDatastore, 5)
	t.Cleanup(cachingBackend.Close)
```

File: pkg/storage/storagewrappers/caching.go [hunk 9]
```diff
@@ -51,15 +51,15 @@ func (c *cachedOpenFGADatastore) ReadAuthorizationModel(ctx context.Context, sto
 	return model, nil
 }
 
-// FindLatestAuthorizationModelID returns the last model `id` written for a store.
-func (c *cachedOpenFGADatastore) FindLatestAuthorizationModelID(ctx context.Context, storeID string) (string, error) {
-	v, err, _ := c.lookupGroup.Do(fmt.Sprintf("FindLatestAuthorizationModelID:%s", storeID), func() (interface{}, error) {
-		return c.OpenFGADatastore.FindLatestAuthorizationModelID(ctx, storeID)
+// FindLatestAuthorizationModel returns the last model `id` written for a store.
```
Comment (miparnisari) [inline pkg/storage/storagewrappers/caching.go:54]:
```suggestion
// FindLatestAuthorizationModel see [storage.AuthorizationModelReadBackend].FindLatestAuthorizationModel.
```

File: pkg/storage/storagewrappers/caching_test.go (hunk 7 above)
Comment (adriantam) [reply pkg/storage/storagewrappers/caching_test.go:59]: added back

File: pkg/storage/storagewrappers/caching_test.go (hunk 4 above)
Comment (adriantam) [reply pkg/storage/storagewrappers/caching_test.go:69]: updated

**Review by miparnisari — DISMISSED** [review]

File: pkg/typesystem/resolver.go [hunk 10]
```diff
@@ -45,37 +45,38 @@ func MemoizedTypesystemResolverFunc(datastore storage.AuthorizationModelReadBack
...
 
-				return nil, fmt.Errorf("failed to FindLatestAuthorizationModelID: %w", err)
+				return nil, fmt.Errorf("failed to FindLatestAuthorizationModel: %w", err)
 			}
 
-			modelID = v.(string)
-		}
-
-		key := fmt.Sprintf("%s/%s", storeID, modelID)
+			key = fmt.Sprintf("%s/%s", storeID, v.(*openfgav1.AuthorizationModel).GetId())
```
Comment (jon-whit) [inline pkg/typesystem/resolver.go:62]: nit (non-blocking): it'd be nice to separate the type assertion from the getter.

```go
model := v.(*openfgav1.AuthorizationModel)

key = fmt.Sprintf("%s/%s", storeID, model.GetId())
```

**Review by jon-whit — APPROVED** [review]

**Review by jon-whit — APPROVED** [review]

**Pull request merged**